src/network/ledger.rs

```diff
@@ -857,177 +878,116 @@ impl<N: Network, E: Environment> Ledger<N, E> {
-            let mut maximum_common_ancestor = 0;
-            // Determine the first locator (smallest height) that does not exist in this ledger.
-            let mut first_deviating_locator = None;
-
-            // Verify the integrity of the block hashes sent by the peer.
-            for (block_height, (block_hash, _)) in maximum_block_locators.iter() {
-                // Ensure the block hash corresponds with the block height, if the block hash exists in this ledger.
-                if let Ok(expected_block_height) = self.canon.get_block_height(block_hash) {
-                    if expected_block_height != *block_height {
-                        let error = format!("Invalid block height {} for block hash {}", expected_block_height, block_hash);
-                        trace!("{}", error);
-                        self.add_failure(peer_ip, error).await;
-                        return;
-                    } else {
-                        // Update the common ancestor, as this block hash exists in this ledger.
-                        if expected_block_height > maximum_common_ancestor {
-                            maximum_common_ancestor = expected_block_height;
-                        }
-                    }
-                } else {
-                    // Update the first deviating locator.
-                    match first_deviating_locator {
-                        None => first_deviating_locator = Some(block_height),
-                        Some(saved_height) => {
-                            if block_height < saved_height {
-                                first_deviating_locator = Some(block_height);
-                            }
-                        }
+        // Process peer information and determine what requests to send.
+        let (start_block_height, end_block_height, requires_revert) = match handle_block_requests::<N, E>(
+            latest_block_height,
+            latest_cumulative_weight,
+            maximal_peer,
+            maximal_peer_is_fork,
+            maximum_block_height,
+            maximum_cumulative_weight,
+            maximum_common_ancestor,
+            first_deviating_locator,
+        )
+        .await
+        {
+            Some(result) => result,
+            None => {
+                // Case 2(c)(b)(a) - Check if the real common ancestor is NOT within `MAXIMUM_FORK_DEPTH`.
+                // If this peer is outside of the fork range of this ledger, proceed to disconnect from the peer.
+                if let Some(first_deviating_locator) = first_deviating_locator {
+                    if latest_block_height.saturating_sub(*first_deviating_locator) >= E::MAXIMUM_FORK_DEPTH {
+                        debug!(
+                            "Peer {} has exceeded the permitted fork range of the protocol, disconnecting",
+                            peer_ip
+                        );
+                        self.disconnect(peer_ip, "exceeded fork range").await;
```

Comment:
there used to be a `return` right away below the disconnect, but it seems that now it's only 6 lines later, after 2 scopes are exited